### PR TITLE
Feature/Bolt Carousel

### PR DESCRIPTION
**Pull request by rockymountainhigh1943**
## Jira

http://vjira2:8080/browse/BDS-73

## Summary

Added the ability for Bolt to provide extendable carousel functionality.

## Details

This PR includes all the initial work for implementing the new bolt carousel. As it is a work in progress, there are still a number of issues to resolve:
- The carousel doesn't appear to fully initialize until after a window event, such as a window resize
- Importing the Swiper styles isn't working (wondering if this has to do with Swiper not coming with .scss to import, but instead .less or .css) so I had to add them manually (yuck) to the carousel .scss to keep making progress
- When the shadowDOM is enabled, the JS doesn't seem to fire and we're missing nearly all Swiper styles (probably related to the issue above)

## How to test

1. Pull down branch, from the root run `yarn install && npm run setup`
1. `cd` to apps/pattern-lab and run `npm run start`
1. Once PL has launched – navigate to http://localhost:3000/pattern-lab/?p=viewall-components-carousel to view demos and the above issues

### CODE REVIEW

**Comment by sghoweri** [comment]
@rockymountainhigh1943 nice! Here’s some initial feedback to get you going!

- [ ] The carousel currently doesn’t work without JavaScript — you can’t scroll ☹️. Progressively enhancing the initial markup first (before focusing on the JS) is SUPER important since this is the foundation your JS is going to be building off of
- [ ] Don’t show arrows till JS kicks in
- [ ] Make sure the far left arrow gets some opacity when you can’t swipe any more to go left / the right arrow gets some opacity when you reach the end 
- [ ] When the JS initially kicks in (without resizing), scrolling is super broken on mobile
- [ ] Nix the slides attribute from the Web Component’s defined props (keeping the “slides” prop in Twig is fine) — the only way we should be able to pass in content into the Web Component should be via composition.

```
<bolt-carousel>
	<bolt-carousel-item>...</bolt-carousel-item>
	<bolt-carousel-item>...</bolt-carousel-item>
	<bolt-carousel-item>...</bolt-carousel-item>
</bolt-carousel>
```

- [ ] Test current implementation’s behavior when using w/ only custom elements (not just via Twig includes)
- [ ] Hide arrows when all slides are visible
- [ ] Animation right now seems very jumpy, even when only having just one single carousel on the screen. Are these animating via 3d transforms?
- [ ] looping is always on — if anything it should be always off by default (and maybe have a prop to switch on)
- [ ] Q. How is each slide getting wrapped in a slide wrapper right now? ie the `sniper-slide` CSS class
- [ ] Remove vendor prefixes from imported Swiper Sass
- [ ] Don’t use the built-in Swiper icons in the Sass — remove and use embedded Bolt SVG icons instead
- [ ] Each carousel slide should automatically center the content vertically and horizontally (my gut tells me we shouldn’t need any alignment options with that) 
- [ ] Make sure we have two specific JS files — one for the main carousel itself and another for the individual carousel slides

## For Reference
Check out the Ionic Framework’s Swiper implementation for some inspiration (spoiler: they also using Web Components via StencilJS)
	- https://github.com/ionic-team/ionic/tree/master/core/src/components/slides
	- https://github.com/ionic-team/ionic/tree/master/core/src/components/slide

**Comment by rockymountainhigh1943** [comment]
Hi @sghoweri - I've gone through and added the last little adjustments and fixes from the PR feedback. 

While adding the vertical centering was fairly simple, I wasn't able to come up with a simple solution for horizontal centering without using text-align, which is obviously problematic. Let me know if we really need that in place and if you have some thoughts on implementation. Otherwise I would say this is good to go!

Thanks!
Jake

File: packages/components/bolt-carousel/carousel.schema.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+title: 'Bolt Carousel'
+description: 'Carousels provide a mechanism for swiping through a collection of content items or components.'
+type: object
+properties:
+  loop:
+    type: boolean
+    title: 'Loop carousel'
+    description: '`True` to enable infinite loop.'
+    default: false
+  autoplay:
```
Comment (mikemai2awesome) [inline packages/components/bolt-carousel/carousel.schema.yml]: There needs to be a more complex `transition` prop that covers both loop and autoplay, for example:

```yml
transition:
  type: object
  properties:
    slide_duration:
      type: string
      default: 3
    transition_duration:
      type: string
      default: 0.5
    autoplay:
      type: boolean
      default: false
    loop:
      type: boolean
      default: true
```

slide_duration: how long does a slide stop for
transition_duration: how long does it take to transition one slide to the next

File: packages/components/bolt-carousel/carousel.schema.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
+$schema: 'http://json-schema.org/draft-04/schema#'
```
Comment (mikemai2awesome) [inline packages/components/bolt-carousel/carousel.schema.yml]: This is lacking props for control, for example:

```yml
controls:
  type: object
  properties:
    swipe:
      type: boolean
      default: true
    progress_indicator:
      type: boolean
      default: true
    navigation:
      type: boolean
      default: false
```

File: packages/components/bolt-carousel/carousel.scss [hunk 3]
```diff
@@ -0,0 +1,642 @@
...
+ * Most modern mobile touch slider and framework with hardware accelerated transitions
+ * http://www.idangero.us/swiper/
+ *
+ * Copyright 2014-2018 Vladimir Kharlampidi
+ *
+ * Released under the MIT License
+ *
+ * Released on: July 31, 2018
+ */
+.swiper-container {
```
Comment (mikemai2awesome) [inline packages/components/bolt-carousel/carousel.scss]: None of this is following our BEM conventions.

File: packages/components/bolt-carousel/carousel.schema.yml (hunk 2 above)
Comment (mikemai2awesome) [reply packages/components/bolt-carousel/carousel.schema.yml]: Also missing a `slides` prop instructing what kind of data can be passed in what structure.

**Comment by sghoweri** [comment]
Closing this one out - reopening a new PR once updates are in place

**Pull request closed without merging**